src/main/java/com/mongodb/hibernate/internal/type/MongoStructJdbcType.java

```diff
@@ -0,0 +1,218 @@
+import static com.mongodb.hibernate.internal.type.ValueConverter.toBsonValue;
+import static com.mongodb.hibernate.internal.type.ValueConverter.toDomainValue;
+
+import com.mongodb.hibernate.internal.FeatureNotSupportedException;
+import java.io.Serial;
+import java.sql.CallableStatement;
+import java.sql.JDBCType;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.hibernate.annotations.Struct;
+import org.hibernate.metamodel.mapping.EmbeddableMappingType;
+import org.hibernate.metamodel.spi.RuntimeModelCreationContext;
+import org.hibernate.type.descriptor.ValueBinder;
+import org.hibernate.type.descriptor.ValueExtractor;
+import org.hibernate.type.descriptor.WrapperOptions;
+import org.hibernate.type.descriptor.java.JavaType;
+import org.hibernate.type.descriptor.jdbc.AggregateJdbcType;
+import org.hibernate.type.descriptor.jdbc.BasicBinder;
+import org.hibernate.type.descriptor.jdbc.BasicExtractor;
+import org.hibernate.type.descriptor.jdbc.StructJdbcType;
+import org.jspecify.annotations.Nullable;
+
+/** Thread-safe. */
+public final class MongoStructJdbcType implements StructJdbcType {
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    public static final MongoStructJdbcType INSTANCE = new MongoStructJdbcType();
+    private static final JDBCType JDBC_TYPE = JDBCType.STRUCT;
+
+    private final @Nullable EmbeddableMappingType embeddableMappingType;
+
+    private final @Nullable String structTypeName;
+
+    private MongoStructJdbcType() {
+        this(null, null);
+    }
+
+    private MongoStructJdbcType(
+            @Nullable EmbeddableMappingType embeddableMappingType, @Nullable String structTypeName) {
+        this.embeddableMappingType = embeddableMappingType;
+        this.structTypeName = structTypeName;
+    }
+
+    @Override
+    public String getStructTypeName() {
+        return assertNotNull(structTypeName);
+    }
+
+    /**
+     * This method may be called multiple times with equal {@code sqlType} and different {@code mappingType}.
+     *
+     * @param sqlType The {@link Struct#name()}.
+     */
+    @Override
+    public AggregateJdbcType resolveAggregateJdbcType(
+            EmbeddableMappingType mappingType, String sqlType, RuntimeModelCreationContext creationContext) {
+        return new MongoStructJdbcType(mappingType, sqlType);
+    }
+
+    @Override
+    public EmbeddableMappingType getEmbeddableMappingType() {
+        return assertNotNull(embeddableMappingType);
+    }
+
+    @Override
+    public BsonDocument createJdbcValue(Object domainValue, WrapperOptions options) throws SQLException {
+        var embeddableMappingType = assertNotNull(this.embeddableMappingType);
+        if (embeddableMappingType.isPolymorphic()) {
+            throw new FeatureNotSupportedException("Polymorphic mapping is not supported");
```

Comment:
Do we plan to support it in the future? Seems the decision is not based on "there is no MongoDB counterpart, so we decided to drop the support tottally). If yes, should we need to create a ticket to avoid loosing track of it? Ideally we also could create testing case for such `FeatureNotSupportedException` as @vbabanin emphasized (and I adopted since his participation of Hibernate PR review).

We did create a related ticket at https://jira.mongodb.org/browse/HIBERNATE-69, though not sure whether we need to reuse it or create a new one.

Reply:
I included embeddables in [HIBERNATE-69](https://jira.mongodb.org/browse/HIBERNATE-69).

Done in 48d5171c19d805b0edbd69c781f91cc04f97ec67.